### ISSUE TITLE
[UX] Merge all the "Update core/module/theme/layout" pages into a single page.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2714]
## Describe your issue or idea

Currently, in the "Available Updates report" page, the site admin can see which projects have updates, but the actions they can do are limited to a) downloading the .zip file of an update and b) visit the release notes page of that update.

In the meantime, we have the issue where the respective "Update xyz" pages for modules (`admin/modules/update`), themes (`admin/appearance/update`) and layouts (`admin/structure/layouts/update`) show all available updates of all project types **plus core**. So, I fail to see how they belong to the sections we currently have them in. Notice for example in the screenshot below how in the "Update themes" page, there is an update for a module as well as an update for core listed:

![image](https://cloud.githubusercontent.com/assets/2423362/26524927/cdc19c3c-4389-11e7-8539-62b7046536c9.png)

True, we could filter the results in those pages by project type, and the UX WTF will be gone. But I think that we have better options...

In the same screenshot, I like how we have the "Download these updates" button there, but we also have the installation queue in the Project Browser. Instead of this screen where the user needs to click the checkboxes of available updates, they could use the "Add to installation queue" links on the Project Browser and then go through the familiar process of the Project Installer. All we'd have to do is add "Up to date", "Security update required!" and "Update available" to the existing "Already enabled" and "Already downloaded" statuses of installed projects. Rough mockup of how this could look:

![image](https://cloud.githubusercontent.com/assets/2423362/26532120/709df5f6-443c-11e7-9fcb-76b18cf0d023.png)

Adding to that, the "Includes: [enabled]/[disabled]" information that we have in the "Available updates" page for installed projects can be appended to the project descriptions in the Project Browser. We could finally have the entries in the Project Browser be color-coded (green/orange/red) based on update status.

### Steps to reproduce/Actual behavior

1. Intentionally install and enable an old version of a module.
2. Go to the "Available updates" page -> you can see that there are updates available, but you cannot install them from that page.
3. Go to `admin/appearance/update` -> the available update is listed in the "Update themes" page although it is a module update.

### Expected behavior (if reporting a bug)

1. The user visits a single page where all projects are listed (both installed and new).
2. The user can filter by an "Updates available" filer, that is only made available if there are updates for installed projects.
3. The user can filter by a "Project type" filter.
4. The user can filter by installed/new projects.
5. The user can use the "Add to installation queue" link to add the module to the queue.
6. The old "Available updates" link in the admin bar redirects to the Project Browser page with the "Updates available" filter passed on as a parameter.

---

Advocate: @klonos 

---

PR: https://github.com/backdrop/backdrop/pull/3451

**Comment by olafgrabienski** [comment]
@klonos I see your point! Looks like a bigger undertaking which needs longer discussion though. What do you think about branching off one baby step to fix the immediate UX issue more quickly, adding for now a simple link from the *Available updates* page to one of the *Update Xyz* pages? (Which one, doesn't matter at the moment.)

**Comment by klonos** [comment]
This is one of those "told you so" issues 😛 which I am bringing back to the 1.x milestone...

This was brought up during last week's dev meeting. What triggered this was the recent core security update + the fact that we have implemented core updates from the admin UI in 1.11.0 (disabled by default), and that we are planning to enable this UI by default in the upcoming 1.12.0 release.

To recap, the idea here is to merge all the "Update xyz" pages (for modules/themes/layouts) into a single "Update all" page. A possibility discussed was to use the "Available updates" page for that purpose. In other words, allow people to perform updates directly from the report page, and then remove the individual "Update xyz" pages that we currently have in place.

This makes more sense now IMO, since we are planning to use the "Settings" page under `/admin/reports/updates/settings` for the automatic security updates UI/options (#414). As @jenlampton mentioned in the dev meeting this past week, this was discussed in very early stages of Backdrop, after the Project Browser got implemented. Back then, people ("people" could have been mainly me 😛) were complaining about too many pages where we can perform updates. She continued to say:

> ...we said we'll just leave it alone, and figure it out later."

...well, it seems that "later" is now 😄

**Comment by klonos** [comment]
...actually, lets see if we can get this in with 1.12, because it would be an awesome new year release feature! 😄

**Comment by olafgrabienski** [comment]
> To recap, the idea here is to merge all the "Update xyz" pages (for modules/themes/layouts) into a single "Update all" page. A possibility discussed was to use the "Available updates" page for that purpose. In other words, allow people to perform updates directly from the report page, and then remove the individual "Update xyz" pages that we currently have in place.

I agree! The current places of the update options are quite far away when a person is on the report pages. Just had the very problem: After enabling the core update for a site via the respective settings file, I expected the option to update core somewhere around `admin/reports/updates`. When I didn't find it, I thought, something with the settings file must have been wrong. Only at a certain point, I had a look at the *Update Modules* page and was able to continue with the update.

**Comment by klonos** [comment]
...yep, that + the fact that although the individual updates pages are meant to be for each project type respectively, they include *all* types + core, which makes it a confusing UI (I had reported this as a bug in #2323).

**Comment by jenlampton** [comment]
There's no PR here yet, so bumping milestone to 1.13. I do think this should be a priority though, since more people are going to start using the updater now, getting this sorted soon would be ideal 👍 

> A possibility discussed was to use the "Available updates" page for that purpose.

I dislike this proposal because I think it will clutter the Available Updates report. That report will be used by people who are not doing automatic or manual updates, and still needs to perform the duties it does now. 

I think it would make more sense that the Installer module would extend the Available Updates report by adding an action link at the top to perform updates (when necessary) and that action link should take people to a single new page (owned by the Installer module, instead of by System module) that consolidates the the separated modules/theme/layout updates pages we have now. Any of those available updates can be performed there.

For simplicity's sake, I think we should start by doing this single consolidation, and getting that working. We can then use follow-up issues to change the update UX, to more closely match what @klonos describes earlier in this thread.

**Comment by jenlampton** [comment]
From @laryn in https://github.com/backdrop/backdrop-issues/issues/3039#issuecomment-382804979:

> I like the direction towards simplifying. Stepping one step even further back, what about when you click directly on "Functionality" and get the list of modules (enabled or not). If there are updates, it's noted in a notification at the top of the page. Would it make sense on this top level page to do:

* subtle indicators of some kind to show which are up to date, which are outdated, etc.
* some kind of sorting here like was done on the status page, so security updates are at the top, followed by other available updates, followed by the usual list? (or at least a way to filter to show only the ones that have updates pending in a quick sort kind of way)
* links to update a module directly from this page?

**Comment by jenlampton** [comment]
> * subtle indicators of some kind to show which are up to date, which are outdated, etc.

yes please, I love this idea! 

> sorting here like was done on the status page, so security updates are at the top...

no, not here, because that's not what this page is for. On the updates page though, yes please!

> links to update a module directly from this page?

How about links to the update page, perhaps with that module pre-checked in the list to update?

**Comment by jenlampton** [comment]
from @laryn in https://github.com/backdrop/backdrop-issues/issues/3039#issuecomment-390327191 

Mockup of "Available Updates" for discussion, adding an "Update all" button at top, and individual "Update" buttons by the individual entries. (Visual includes [sort changes](https://github.com/backdrop/backdrop-issues/issues/3132) and [ability to update core](https://github.com/backdrop/backdrop-issues/issues/3105))

![screen shot 2018-05-18 at 3 40 48 pm](https://user-images.githubusercontent.com/871421/40256999-1b40cab0-5ab2-11e8-8844-6177040c3f69.jpg)

I'm also thinking that maybe the ones with no updates don't even need to show on this page, to be honest.

**Comment by jenlampton** [comment]
> Mockup of "Available Updates" for discussion

I dislike this proposal because I think it will clutter the Available Updates report. That report will be used by people who are not doing automatic or manual updates, or don't have the Installer module enabled. It still needs to perform the duties it does now. 

I think it would make more sense that the Installer module would extend the Available Updates report by adding an action link at the top to perform updates (when necessary) and that action link should take people to a single new page (owned by the Installer module, instead of by System module) that consolidates the the separated modules/theme/layout updates pages we have now. Any of those available updates can be performed there.

For simplicity's sake, I think we should start by doing this single consolidation, and getting that working. We can then use follow-up issues to change the update UX, to more closely match what @klonos describes earlier in this thread.

**Comment by jenlampton** [comment]
This is a huge UI change, and with less than a week left before code freeze, it's not looking likely that it will get in. Bumping milestone to 1.14.

**Comment by jenlampton** [comment]
Removing milestone, adding milestone candidate label. If this issue gets an advocate who wants to push it through the 1.14 release, it can get the milestone back :)

**Comment by quicksketch** [comment]
I think the mockup in https://github.com/backdrop/backdrop-issues/issues/2714#issuecomment-450793020 is actually pretty nice. I always get confused between there being multiple pages that show me which updates are available and which ones will actually allow me to do the updating. If there's only one page to do everything it makes linking to a page that can actually get things done easier. For example when there's a security update, we link to the available updates page. But you can't actually do anything there. You then have to go the modules update page to actually apply the updates.

> That report will be used by people who are not doing automatic or manual updates, or don't have the Installer module enabled.

We could hide the update buttons in those cases.

---

One thing I can't figure out if we did make a new unified "All available updates" page: where would it live in the navigation? Under "Functionality" seems okay, but the URL for that is actually /modules. So Functionality -> Updates would actually be `/modules/updates`.

**Comment by olafgrabienski** [comment]
> ... if we did make a new unified "All available updates" page: where would it live in the navigation? Under "Functionality" seems okay, but the URL for that is actually /modules. So Functionality -> Updates would actually be `/modules/updates`.

Indeed! When the menu item "Modules" was renamed to "Functionality" in #209, the path wasn't changed respectively. So, currently we have the page *Update modules* (`admin/modules/update`, without an `s`). There, on the current *Update modules* page, people can already update every type of project (modules, Backdrop core, themes), and the same happens on the *Update themes* page (`admin/appearance/update`).

Compared to the current situation, in my opinion a single page with the same title and same path for all kind of projects would already be an improvement, even if the path would contain `modules`. (I guess, discussing path changes is still quite complex, so that it would block an important improvement here.)

**Comment by olafgrabienski** [comment]
Additional idea: add a quick link to the "Home" menu item which points to the new, unified project update page.

There is already a "Run updates" link, which points to `update.php`. I'd suggest to rename this item in any case to "Run database updates" or similar.

**Comment by klonos** [comment]
Yup, we can link to that unified page from the new Dashboard, and the "Available updates" block 😉

For reference, in the latest D8 dev version, they have added a "+ Install new module or theme" button in the "Available updates" page:

<img width="667" alt="Screen Shot 2019-06-30 at 8 41 54 am" src="https://user-images.githubusercontent.com/2423362/60390133-f8f47e00-9b12-11e9-9af1-599f06e4005f.png">

**Comment by klonos** [comment]
Here's another quirk...

I understand the value of denoting a theme in the "Update modules" page:
![image](https://user-images.githubusercontent.com/2423362/94187198-498e8e80-feb0-11ea-904d-435bce07e94d.png)

...but why aren't layout templates denoted?

Also, when you visit the "Update themes" page, the theme is still denoted:

![image](https://user-images.githubusercontent.com/2423362/94187500-b4d86080-feb0-11ea-9022-8753e36ab9a0.png)

I want to get this in for 1.18.0.

**Comment by BWPanda** [comment]
I don't want to undo any discussions or decisions that have already happened, and I'm sorry to come late to the party, but I just listened to the discussion in the [weekly meeting](https://www.youtube.com/watch?v=vVwy8UOZoD0) and have some thoughts/clarifications...

If I understand it correctly, this is the current situation:
- `admin/appearance` (manage installed themes)
- `admin/appearance/install` (install new themes)
- `admin/appearance/update` (update installed themes (and apparently other projects))
- `admin/modules` (manage installed modules)
- `admin/modules/install` (install new modules)
- `admin/modules/update` (update installed modules (and apparently other projects))
- `admin/structure/layouts/settings` (manage installed layout templates)
- `admin/structure/layouts/install` (install new layout templates)
- `admin/structure/layouts/update` (update installed layout templates (and apparently other projects))
- `admin/reports/updates` (list installed projects)

The problem we're trying to address is the '(and apparently other projects)' parts, right? In that the update page for themes shouldn't also list updates for core, modules, etc.

And this is the new proposal:
- `admin/appearance` (manage installed themes)
- `admin/appearance/install` (install new themes)
- `admin/modules` (manage installed modules)
- `admin/modules/install` (install new modules)
- `admin/structure/layouts/settings` (manage installed layout templates)
- `admin/structure/layouts/install` (install new layout templates)
- `admin/reports/updates` (list installed projects)
- `???` (update installed projects)

With the current situation, we have effectively 4 'areas': `admin/appearance`, `admin/modules`, `admin/structure/layouts` and `admin/reports/updates`. The first 3 all relate to the three main project types for Backdrop, while the 4th is a list of all projects in a Backdrop site.

The new proposal seems to be wanting/needing to make a 5th area (URL to be determined). This is because it can't be added to one of the first 3 areas (since they're all project-type-specific), and, as @jenlampton mentioned in the meeting, it can't be added to the 4th area (since that's a report and just lists information; it doesn't/shouldn't let you actually *do* anything). So we need a 5th area for updating projects.

To me, this seems to be adding more, unnecessary complexity; which, dare I say, might make it harder for users than the current situation.

I suggest doing one of the following:
- Just remove the projects from each of the 3 'update' pages that shouldn't be there (e.g. `admin/appearance/update` should just list themes, so remove core, modules, etc.).
- Add the ability to update projects to their 'manage' pages (e.g. `admin/appearance` lets you manage *and update* installed themes).

Wordpress seems to do the second option (based on the following screenshot I made from [this video](https://wordpress.org/support/article/plugins-themes-auto-updates/#bulk-enabledisable-plugins-auto-updates)):
![image](https://user-images.githubusercontent.com/2385329/96947259-6d380980-152e-11eb-878b-489b29175b72.png)

**Comment by indigoxela** [comment]
@BWPanda many thanks for chiming in.

> The new proposal seems to be wanting/needing to make a 5th area (URL to be determined).

Ehm... not actually. Personally, I'd prefer to have one and only one page, where all updates can be done. Themes, layouts, modules and core. Having many different pages (with different menu items in the admin menu) seems odd and unnecessary. Or is there any possible use-case for that?

The current situation is that I can update core via the themes update page - which I find confusing.

**Comment by BWPanda** [comment]
@indigoxela I think you misunderstand me...

> I'd prefer to have one and only one page, where all updates can be done.

Yes, that seems to be the concensus here. But my problem with that is then where do you put this new page that lists all updates in the one place...? You can't put it in the `admin/appearance` section, because that's only for themes (not for modules, layouts, etc.), and likewise you can't put it in `admin/modules` or `admin/structure/layouts` either. Jen rightly stated in the meeting video (linked above) that the `admin/reports` section is for displaying information, not for performing actions, so it can't go there. That means a new section/URL has to be added for updating projects. That'll mean there'll be 5 different sections of the site where projects are listed/managed/updated/etc. I think that's going to be confusing...

> Having many different pages ... seems odd and unnecessary. Or is there any possible use-case for that?

Well we currently already have different pages for *installing* new modules, themes and layouts (`admin/modules/install`, `admin/appearance/install` and `admin/structure/layouts/install` respectively). Or am I misunderstanding you and you're suggesting we merge all of those too? If so, that'll greatly expand the scope of this issue, but it might be a worthwhile discussion to have.

Since we already have 3 different sections of the site for managing and installing the 3 different types of projects for Backdrop, I'm suggesting keeping the updates in those same sections, rather than creating a new section just for updates.

I hope that's clearer...? :slightly_smiling_face:

**Comment by indigoxela** [comment]
> I think you misunderstand me...

I seem to have a talent for that, yes. :wink:

To clarify: I'm not suggesting to merge all install pages. These are fine and make sense to me exactly as they are.

The most reasonable place for the unified update page, at least in my opinion, is somewhere close to `admin/reports/updates`.
Drupal 7 uses `admin/reports/updates/update`.

> ...I'm suggesting keeping the updates in those same sections

I suppose, you suggest to filter these individual pages for their respective project types?

But that means that I _have_ to go to four different places, spread all over the admin menu, just to do something that could have been done on a single page.

- admin/appearance/update - for themes
- admin/structure/layouts/update - for layouts
- admin/modules/update - for modules

And following your suggestion, we would need a new update page for core, as we don't have that yet.

- admin/core/update (doesn't exist yet)

That doesn't seem like the best solution to me.

**Comment by BWPanda** [comment]
Mmm, I see what you mean. You don't normally install modules and themes and layouts all at once, hence its OK to have them on different pages. But you often want to update everything at the same time, so having them all on the same page fits that requirement. Good point!

**Comment by olafgrabienski** [comment]
> the `admin/reports` section is for displaying information, not for performing actions, so [the update page] can't go there

I agree but we could link from the report to the update page and put that link even in the Reports menu (as we do with `Reports > Top 'page not found' errors > Fix 404 pages ...` which points to `admin/config/urls/redirect/404`).

A good place for the merged update page: what about `admin/config/system`?

Additionally, we could put the update page link in the admin bar "Home" menu, together with "Flush all cashes" ecc.

**Comment by klonos** [comment]
> Just remove the projects from each of the 3 'update' pages that shouldn't be there...

That would be great @BWPanda, but the problem is that we also have core updates; and these are shown in all 3 areas. So even if we filtered these 3 "updates" pages to only show the respective project type, we'd still have a page called "Update modules" which also allows you to update core. Same for the "Update themes" and the "Update layouts" page (which should really be "Update layout templates", but deferring that for another discussion). So what would then be appropriate would be to have the pages be called "Update modules and core", "Update themes and core", and "Update layout templates and core". You can see how that's not ideal (lengthy tab names, and also not practical to have it so that core can be updated in 3 different places). So what do we do? ...do we introduce a separate 4th place, dedicated to only updating core? ...and if so, where should that place live in the admin menu structure?

> That means a new section/URL has to be added for updating projects. That'll mean there'll be 5 different sections of the site where projects are listed/managed/updated/etc. I think that's going to be confusing...

The goal is to add a new place where all things can be updated, **but then also remove the 3 existing places where we can update different types of thingies**. So we'd end up with only 2 places really: one to report updates, and one to perform the updates (I believe that recent consensus is the latter to be under `/configure/system/updates`).

> But you often want to update everything at the same time, so having them all on the same page fits that requirement. Good point!

Exactly! I think that the simplest solution would be to have a single place to view a report about available updates, and then another single place where you can select and perform these updates. If we want to further merge those 2 places into a single one, then a separate issue should be filed (because there's the whole discussion about one thing being a report page, wheres the other is a configuration/action page).

So to recap...

UX issues with the current situation:
- listing updates of all project types in project-type-specific areas seems sloppy/inconsistent and is confusing
- listing core updates in a place that is project-type-specific is also confusing
- if we filtered each update page by its respective project type, then we'd:
  - potentially have people go to 3 different places to perform a single task
  - have core updates either be "homeless" or listed in all 3 places (which is confusing)

So the proposal is to:
- introduce a new single place where updates of all kinds can happen
- deprecate the routes for all 3 existing "update" pages, and redirect to the new "update all thingies" place

**Comment by klonos** [comment]
I plan to start working on this soon, and I'm also advocating for it for 1.18.0.

**Comment by klonos** [comment]
I'm making some good progress with this on my local, here's what the unified "Updates" page looks like (located under: Configuration -> System -> Updates):

![image](https://user-images.githubusercontent.com/2423362/101786743-f4553700-3b06-11eb-911f-ad12af137a36.png)

What's left to be done:

- hide/show the submit buttons using `#states`, based on whether updates have been selected or not
- add a filter at the top, for enabled/disabled, core/contrib, project type, and update type (i.e. only security updates)

PS: there are sooo many inconsistencies and things that can be improved re how things work, that it makes my head hurt 😅 ...I'm trying (hard) to keep away from all that, and keep the changes for this to a minimum.

**Comment by laryn** [comment]
I'm liking this direction 👍

**Comment by klonos** [comment]
I've been "battling" some CSS issues, but it turned out to be a pre-existing problem: #4816

If you notice this happening while testing the PR:

![image](https://user-images.githubusercontent.com/2423362/102676247-4f40fb00-41a5-11eb-8c88-0a2815f90834.png)

...then it's not because of my changes 😅

**Comment by klonos** [comment]
Sorry this has taken this long ...there are a few tiny things to address still, but here's a PR for everyone's pleasure: https://github.com/backdrop/backdrop/pull/3451

- The `BACKDROP_VERSION` version change in `bootstrap.inc` is intentional, in order to test core security updates, and will be removed once I've polished and cleaned up everything.
- You might see the occasional `dpm()` or commented-out code here and there. This will also be cleaned up after I've addressed any feedback.

Please provide feedback, as I would like to take advantage of these couple of next weeks, to drive this to end. Thanks 🙂

**Comment by klonos** [comment]
So thoughts:

- This page seems pretty "dull". Is there anything that we need to add at the top, like some help text or info message?
- Should we move the `/admin/reports/updates/settings` page under `/admin/config/system/updates/settings` as well? If so, do it as part of this issue here, or follow-up instead?

**Comment by BWPanda** [comment]
In testing the PR, I have a slight concern... It seems as if now the Update page is hidden away. Previously, updates were more prominent as tabs on top-level menu items. Now it seems buried in the admin menu.

If other's disagree with me then I'm happy to go with the majority. But it is the first thing I notice with this new page.

**Comment by klonos** [comment]
Thanks for the feedback @BWPanda 🙏 

Well, you are only meant to go to that page IF there are updates; whereas before, all those "Update xyz" tabs were there all the time. Besides, if there are security updates, there is a big red warning at the top of the modules/themes/layouts listings, with a link to the "Available updates" page:

![image](https://user-images.githubusercontent.com/2423362/102677900-78fe2000-41ad-11eb-8c6b-700e2da9274a.png)

Should we change that link to directly point to the new "Updates" page instead of the "Available updates" report?

Also, would it help if we added a warning/info message for regular updates? (but only on the modules/themes/layouts pages - it would otherwise be too much noise for the rest of the admin UI).

**Comment by BWPanda** [comment]
Yes, I think if there are security updates then a red error on all pages pointing to new Updates page. If there are regular updates, a yellow warning only on project listings pointing to new Updates page.

**Comment by klonos** [comment]
OK, I've updated the PR:

- cleaned up some unused code ...mainly removed the bits that were splitting disabled/enabled projects to separate tables - no need for that now (there's a dedicated "Project status" column, and we can add sorting/filtering)
- Replaced the plain-text "No updates are available for any of your projects." with a proper success "Your site is up to date." message (people coming to this page will appreciate the "all good" vibe if no updates 😉 )

  ![image](https://user-images.githubusercontent.com/2423362/102683129-6567af00-41d7-11eb-9960-ae9d423a8dd1.png)

**Comment by klonos** [comment]
...moar tweaks:

- run some basic health checks, and show messages/warnings accordingly
- added last update check indication + button to check manually (same as in the "Available updates" page)

![image](https://user-images.githubusercontent.com/2423362/102683844-7e269380-41dc-11eb-874d-a07bd98369c1.png)

![image](https://user-images.githubusercontent.com/2423362/102683876-aadaab00-41dc-11eb-8e17-24eccf3e5386.png)

**Comment by indigoxela** [comment]
I did a quick test in the sandbox. Additionally to the fake core update status, I also added an outdated version of a module and a theme.

I really love this simplified UI! The workflow seems very smooth and clear to me. I'll to do more local testing later.

**Comment by klonos** [comment]
Here's a list of specific things for which I need feedback (on top of anything else anyone has to suggest):

- Should the updates settings page be moved from `/admin/reports/updates/settings` under `/admin/config/system/updates/settings`? (I think so)

- @BWPanda suggested that it is more tedious to get to the updates page now when you are in any of the modules/themes/layouts pages, as the new "Updates" page is on a different admin bar menu; whereas before, you were one click away, since each update page was a primary navigation tab.

  When you are in any of the modules/themes/layouts pages, and there is a security update for core/contrib, you see an red error:

  ![image](https://user-images.githubusercontent.com/2423362/102677900-78fe2000-41ad-11eb-8c6b-700e2da9274a.png)

  If the available updates are regular updates, you get a warning instead:

  ![image](https://user-images.githubusercontent.com/2423362/102698068-6c2c0b80-4243-11eb-8520-02995b5f59d9.png)

  The second part of these errors/warnings is the same "See the [available updates]() page for more information.", but it links to `/admin/reports/updates`. Should we reword that and link to `/admin/config/system/updates` instead?

  Another option would be to linkify the "update immediately!" (security updates) and "update as soon as possible" (regular updates) bits, and have them link to `/admin/config/system/updates`.

- Should we be showing an error/warning for core + any contrib update (current behavior in the PR), or show these warnings specifically if there are updates of the same type of contrib project instead? In other words show the messages with the links to the new "Updates" page on the modules listing page _only_ if there are _module_ updates, on the themes listing page _only_ if there are _theme_ updates, and on the layout templates listing page _only_ if there are _layout_ updates. Follow up perhaps?

- Should we add a link from `/admin/reports/updates` to allow people to easily get to `/admin/config/system/updates`? (noting that all the "Update available" links are now pointing there) ...also, follow up perhaps?

**Comment by indigoxela** [comment]
> Should the updates settings page be moved from /admin/reports/updates/settings under /admin/config/system/updates/settings?

Hm. We don't have a dedicated "updates" section in the admin menu, so both, "reports" and "config" might not be perfect. I'm fine with any of it. Or should we create an extra section?

> ...it is more tedious to get to the updates page now when you are in any of the modules/themes/layouts pages

Is it really more tedious than it was before? I'll need to play around with the PR a bit more to figure out.

One thing to consider: to _see_ that there are updates available (reports) and to actually update seems to have different permissions.

- access site reports

And probably these overlap:

- administer software updates
- use installer browser
- administer modules
- administer themes

I find that a bit confusing, so I'm not sure yet, from where we should link to where.

Regarding the messages shown on specific pages (when updates are available)... If there are updates available, I'd be fine with a *short* generic message that informs me and provides a link to more details. But that's just my personal preference.

**Comment by laryn** [comment]
Can't test right now but scanned the thread and it is sounding great. Should there be a link to this page from the dashboard "updates" block?

**Comment by klonos** [comment]
> Should there be a link to this page from the dashboard "updates" block?

There is 😉 

![image](https://user-images.githubusercontent.com/2423362/102735877-31fa6100-434c-11eb-8603-d518de83f79e.png)

The "version x.y.z available" link goes to `/admin/reports/updates`, whereas the "Update" link links to `/admin/config/system/updates` ...come to think about it, perhaps we should have a single "Update" link in the bottom-right corner of that block, to brink consistency with the rest if the Dashboard blocks. Previously, each entry on that block would its "Update" link to point to different pages, depending on whether it was a module/theme/layout. Now we don't need that anymore. I'll fix that 👍

**Comment by klonos** [comment]
...OK, I've tweaked the Dashboard updates block. Here's before/after screenshots from my local:

![image](https://user-images.githubusercontent.com/2423362/102737535-58ba9680-4350-11eb-8923-9cf7d3e41312.png)

![image](https://user-images.githubusercontent.com/2423362/102741178-54937680-435a-11eb-88af-63403b2f6976.png)

I still don't like that the "version x.y.z available" links all link to the same thing 🤔

**Comment by indigoxela** [comment]
> I've tweaked the Dashboard updates block. Here's before/after screenshots from my local

Looks good! Any chance that you can push your latest dev state to Github? I'd like to test, how this table works on small screens.

> I still don't like that the "version x.y.z available" links all link to the same thing

Maybe skip the inline links and add only two links at the bottom (or above the table)? One leads to the report page, one leads to the update page (if user has permission).

**Comment by klonos** [comment]
> Any chance that you can push your latest dev state to Github?

Done 🙂 

Tweaked a bit more... no updates:

<img src="https://user-images.githubusercontent.com/2423362/102902257-6cace800-4477-11eb-8adb-72ce2fb8097d.png" width="50%" height="50%">

...updates available:

<img src="https://user-images.githubusercontent.com/2423362/102903042-75ea8480-4478-11eb-8868-88097e639d54.png" width="50%" height="50%">

**Comment by klonos** [comment]
> Maybe skip the inline links and add only two links at the bottom (or above the table)? One leads to the report page, one leads to the update page (if user has permission).

Yeah, I thought of that too, but it'd require updating `dashboard-panel.tpl.php` to support more than one "more" link. That would affect the style of all dashboard blocks though. OK to do as part of this issue, or follow-up perhaps?

**Comment by indigoxela** [comment]
> OK to do as part of this issue, or follow-up perhaps?

Yes, that's absolutely OK. The change is for sure big enough. :wink: 

Something's weird with the dashboard, BTW. I played around in the sandbox (installing and uninstalling modules) and all of a sudden the infos in that dashboard table were gone (empty rows):

![dashboard-table](https://user-images.githubusercontent.com/842066/102905222-150c7d80-4473-11eb-941b-21d6b6a8895d.png)

And I had the same php notices as the failing tests. After hitting the "Check manually" button, the info was back.

Hm, I could just reproduce that again: Install an outdated module, run cron, visit dashboard - OK. Disable that outdated module, visit dashboard - notices, notices, notices...

```
 Notice: Undefined index: title in DashboardUpdateBlock->getContent()
 Notice: Undefined variable: update in DashboardUpdateBlock->getContent()
 ...
```

**Comment by klonos** [comment]
Yeah, I saw these failures too now, but it worked w/o an issues on my local. ...I'll need to revisit/check the logic some more. Please bear with me.

**Comment by indigoxela** [comment]
Regarding the overall workflow (@BWPanda mentioned his concern that the update page might be hidden away):

It doesn't appear like that to me. If there are (security) updates available, the messages haunt you around anyway. And there's also the red circle in the admin bar. And the links from the status page and the dashboard... No chance to miss available update IMO.

Regarding the new menu item location: Configuration / System / Updates seems OK to me - that way it's next to Backup and Migrate if that module's installed.

**Comment by indigoxela** [comment]
Regarding the dashboard block problem:
I verified that current installs do not show any notices on the dashboard page, after a project with available updates has been disabled. My suggestion would be to revert the dashboard (update block) changes as far as possible - only update the links to the unified update page, but stick with the list display for now. We can change the dashboard display in a follow-up.

That gives us more time to figure out, why the switch to the table display is problematic. I do have a suspicion but no definitive root cause yet.

**Comment by klonos** [comment]
> My suggestion would be to revert the dashboard (update block) changes as far as possible - only update the links to the unified update page, but stick with the list display for now. We can change the dashboard display in a follow-up.

I agree 👍 ...in the middle of a Xmas/NY holiday laziness at the moment 😅 , but I'll get to it soon.

**Comment by philsward** [comment]
Thoughts on having each item with its own update link to allow individual updates?

@klonos in the table image you provided, how about moving the update column header to its own update column and replace that with "Version"?

I have scenarios where I don't want a handful of modules to update. Photoswipe beta6 on Drupal is a great example, it's broken.

Writing this makes me realize, a checkbox of what to update would be even better. A person can check all, then uncheck the 1 or 2 they don't want updated.

**Comment by klonos** [comment]
>Thoughts on having each item with its own update link to allow individual updates? ...a checkbox of what to update would be even better. A person can check all, then uncheck the 1 or 2 they don't want updated.

As it is now in the current PR, once you hit the single "Update" link, you are redirected to the update page, where each module has a checkbox (and there's also a "check all" checkbox). So people can still select individual update(s) to perform. See my previous screenshot in https://github.com/backdrop/backdrop-issues/issues/2714#issuecomment-742567723

> ...how about moving the update column header to its own update column and replace that with "Version"?

That's how things are in the Updates page; there are separate columns for current version and available update. With the dashboard block though we have very limited horizontal space, so it needs to be 2-3 columns max. The intention with the block is to provide a "lightweight" overview, then link to the page with all the required info + action elements (checkboxes and update button to actually perform the update). Makes sense?

> We might have to do some thinking down the road whenever the autoupdate is introduced.

All that is being discussed in #414, and I believe that the intention is to only auto-update security updates by default (but still provide an option for all updates I think).

As for the selective notifications and excluding one or more specific projects (until issues are fixed in a specific version), in D7 there was https://www.drupal.org/project/update_advanced which we can consider porting.

**Comment by indigoxela** [comment]
> Thoughts on having each item with its own update link to allow individual updates?

@philsward isn't that the complete opposite of what we're trying to achieve here - simplification? If I get your idea right, you propose to add an update page *for every single installed module*. Please tell me that I misunderstood it. :smirk: 

> I have scenarios where I don't want a handful of modules to update ... a checkbox of what to update

This is something both, the current solution and the improved unified page, provide anyway. You can always select and deselect projects.

**Comment by philsward** [comment]
@indigoxela you misunderstood 😉

From what @klonos said, I think it's all taken care of.

I was just scared the update was going to be all or nothing.

**Comment by klonos** [comment]
> I was just scared the update was going to be all or nothing.

Yup, all good 👍 ...there are checkboxes to allow selecting one/many/all updates. That's how it's always been - just now on a single page.

I believe that the confusion here stemmed from the screenshot of the "Available updates" block, which has no checkboxes, and a single "Update" link. It all makes sense when one realizes that the block does not allow you to perform the updates directly ...it merely serves as an overview/report, that links to the update page, where the actual updating can happen 🙂

**Comment by quicksketch** [comment]
PR is looking pretty good so far. I tested it out locally and the new update page works as expected. Tests are failing, but that's known at this point. The only particularly weird thing I saw is that the admin bar maintains the links to each update page, even though the tabs are hidden.

![image](https://user-images.githubusercontent.com/100206/103469015-51ae6380-4d14-11eb-954c-5a86354a2580.png)

However, clicking on the "Update modules" page properly redirects as intended from `admin/modules/update` to `admin/config/system/updates`.

**Comment by quicksketch** [comment]
@klonos How are you feeling about this PR? We still have a little time left here for the 1.18.0 preview release, but I feel like it would be wiser to delay this into the next release.

**Comment by philsward** [comment]
My vote: if it's 85%+ there, get it added to 1.1.8.

If it's almost there, let's not loose the momentum, and fix glaring problems in the next few bug releases.

**Comment by indigoxela** [comment]
@klonos we could wait until April 29th, to start working again on this issue - or we start earlier. (I'd prefer earlier :smile: )

Just trying to get your attention back on this. I know that you're having lots of other todos in this queue.

**Comment by klonos** [comment]
I'll get back to this soon, I promise 😄 ...but why did you mention April 29 specifically @indigoxela?

**Comment by indigoxela** [comment]
> why did you mention April 29 specifically

It was just an arbitrary example for "too late" - too close to release date IMO. :wink:

**Comment by indigoxela** [comment]
@klonos I added a comment to the PR many days ago with some hints re the dashboard block. Maybe you'll find the time to resume working on it soon? :crossed_fingers:

**Comment by klonos** [comment]
Thanks @indigoxela 🙏🏼 

Also making this note-to-self here re implementing the new `system_redirect_deprecated_page()` function instead of `backdrop_goto_deprecated()`. See https://docs.backdropcms.org/change-records/system_redirect_deprecated_page-should-be-used-for-deprecated-paths

**Comment by quicksketch** [comment]
I'm going to work on getting the tests passing and any minor changes that would need to be made to get this mergeable. I'd love to still have this completed for the next release, we can iterate on it further as needed.

**Comment by klonos** [comment]
Thanks @quicksketch 🙏🏼 ...please give me 5-10min to push some changes from my local before you start working on this.

**Comment by quicksketch** [comment]
I already pushed up a PR to your branch at https://github.com/klonos/backdrop/pull/6. I can adjust further if we get conflicts between your changes and mine. Thanks!

**Comment by klonos** [comment]
OK, cool 👍🏼 ...I'll merge your changes shortly, since I just saw that my local has many changes that I need to weed through (what happened to me committing/pushing small changes often 🤦🏼 )

**Comment by klonos** [comment]
...I've reviewed and merged your PR @quicksketch. Now waiting for the tests to run.

I then plan to implement @indigoxela's suggestion from https://github.com/backdrop/backdrop/pull/3451#discussion_r571448630 and then manually re-add my changes from my local (too many conflicts to work through otherwise).

**Comment by indigoxela** [comment]
Tests are still failing. Please have another look.

**Comment by jenlampton** [comment]
> Tests are still failing. Please have another look.

I ran the failing tests locally, and they all passed (on PHP7) I'll try them again on php 5.3.

I've added some commits to @klonos's PR that do the following:

* add a button on the top of the 'Available updates' report to the page where updates can be performed.
* Changed link text on dashboard block from `Updates` to `Manage system updates`.
* Changed the page title from `Updates` to `System updates`
* Changed the page description to `Review and perform updates for Backdrop core, modules, themes, and layouts.`
* Changed message type from `success` to `info` on the top of the new update page.
* Included the email addresses in the info message about email notifications.
* Added a `Review settings` link for all `warning` messages (it leads to the settings page under reports).
* Added a new `warning` message when email notifications are disabled: `Email notifications are disabled for for available updates.`
* Changed the text for a missing email address to `No email address has been provided for notifications about available updates.`
* Changed the text for disabled automatic update checks to `Automatic update checks are disabled and will need to be run manually.`

Tests are testing again now: https://github.com/backdrop/backdrop/pull/3614

**Comment by jenlampton** [comment]
Okay, I found one failing test and fixed it...  testing again now. 🤞

**Comment by klonos** [comment]
@jenlampton I really like the various messages you've added for the update checking frequency and email recipients, but the "Review settings" link was being added only to the email message, while it also applies to the frequency message. When everything is OK, and both messages are `info`, that does not present much of an issue (although still odd); but when the frequency message is a warning, you'd end up with duplicated links:

![image](https://user-images.githubusercontent.com/2423362/116837646-f57d2280-abd3-11eb-9240-6a8470a368d6.png)

Since we have added the "Install system updates" link next to the "Check manually" button in `admin/reports/updates`, I thought we should reuse the same pattern in `admin/config/system/updates`. So I've renamed the link to "Configure updates settings" instead, and I've added it next to the "Check manually" button:

![image](https://user-images.githubusercontent.com/2423362/116837904-f1053980-abd4-11eb-9ce1-160cac9fe0ec.png)

Do you think that's better/worse?

**Comment by jenlampton** [comment]
> you'd end up with duplicated links

I don't really mind the duplicated links. (You can see from the code that I assumed the same link would appear multiple times on the page, so generated it only once and shared it across all messages). 

I didn't think it was a good idea to tell people something is wrong (by showing them a warning) without giving them a way to fix that problem along with the warning.  I'd rather not have to guess that a link somewhere else on the page is what I need to fix my problem. 

> Since we have added the "Install system updates" link next to the "Check manually" button in admin/reports/updates, I thought we should reuse the same pattern in admin/config/system/updates

I don't like this...  We're now using a pattern that's intended for one thing for something different.

I'm also not a fan of having different email addresses showing up in a bulleted list. It's easier to scan now, but it takes up too much space on the page this way. This message isn't important enough to warrant a giant blue box like this. Are we trying to make them more prominent? We could try bold or italic if you think that might help?

**Comment by jenlampton** [comment]
> you'd end up with duplicated links

Idea: What if we integrated the link more discreetly, by linking existing text within the message?

**Comment by klonos** [comment]
> I didn't think it was a good idea to tell people something is wrong (by showing them a warning) without giving them a way to fix that problem along with the warning.

Agreed 👍🏼 

> Idea: What if we integrated the link more discreetly, by linking existing text within the message?

Yup, that'd work best 👍🏼

**Comment by klonos** [comment]
...how about this?:

- Automatic update checks [are configured](url) to run @frequency.
- Automatic update checks [are disabled](url) and will need to be run manually.
- [Email notifications](url) are disabled for available updates.
- No email address has been provided for [notifications about available updates](url).
- [Update notifications](url) for this site will be sent to:

**Comment by jenlampton** [comment]
What if we made every link a complete sentence on its own:

- [Automatic update checks are configured to run @frequency](url) .
- [Automatic update checks are disabled](url) and will need to be run manually.
- [Email notifications are disabled](url) for available updates.
- [No email address has been provided](url) for notifications about available updates.
- [Update notifications will be sent](url) to:

**Comment by klonos** [comment]
Yup, lets go with that, and we can follow-up if we want to tweak further 👍🏼

**Comment by quicksketch** [comment]
Separate from the link discussion: Would this new page be better if it were located under "Developer" rather than "System"? Seems like configuring updates is in the same vein as "Maintenance Mode".

**Comment by klonos** [comment]
Well, I think we've discussed that already 😅  ...besides, performing updates is not a developer-only task; site admins/owners may do it as well.

Performing updates via FTP, or composer etc. would feel like a developer task, but doing it via the admin UI feels like a site builder (non-coder) task. As a non-developer or site builder, I would first check under Configuration -> System, rather than under Configuration -> Development.

**Comment by jenlampton** [comment]
> Would this new page be better if it were located under "Developer" rather than "System"?

I think that I would have found the page if it were under developer tools. I looked there. I did not look under System.

> performing updates is not a developer-only task; site admins/owners may do it as well.

True, but anyone who's doing updates should have access to the "Developer" section of the menu. That does't define who the tasks are for -- just the sorts of tasks they are. If you need to use "developer tools" to put the site in maintenance mode, and maintenance mode is recommended for updates, then updates are "developer tools" too, and it does make sense to put those things in the same place.

**Comment by quicksketch** [comment]
Yeah, perhaps separate out that problem. I think Maintenance Mode and Updates should be in the same place, whichever that is. Cron however, should be under Developer (it's currently under System). Seems like System vs. Developer needs revisiting entirely.

**Comment by jenlampton** [comment]
My vote is for eliminating "system" entirely :) but...  not in this issue ;)

**Comment by klonos** [comment]
Agreed 👍🏼 ...follow-up: #5075 🙂

**Comment by quicksketch** [comment]
PR at https://github.com/backdrop/backdrop/pull/3451 is now passing all tests (thanks for your patience with the testbot @klonos!) There are some small things yet to be worked out in follow-ups: minor string changes and possibly moving the menu item, but this seems like it's ready to go. And we managed to include the update to the dashboard block after all.

I'd love to have a second approval, based on the testing of the functionality rather than just reading through the code.

**Comment by jenlampton** [comment]
I gave it another test with themes and layout templates, and it's working great! I also did a code review, RTBC from me!

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/3451 into 1.x for 1.19.0. Yay and we still have almost an hour before it's feature freeze in Hawaii. :stuck_out_tongue_winking_eye: 

Thanks folks for all the work leading up to this and those providing the last minute oomph.

**Comment by indigoxela** [comment]
I also quickly checked an update to current HEAD. Works as expected. :+1: 

On a pre-existing 1.18.2 install I replaced the core by the fresh one and ran update.php (which wasn't necessary, actually).

The outdated update links vanished as they should, /admin/config/system/updates was the only remaining one. Installed an outdated version of a module, visited the dashboard. Hit the refresh button ("Check manually") - update shows up, disabled that module and went directly back to the dashboard. No notices, nothing in dblog that shouldn't be there. :+1:

Some wording / messages could be improved, but that are follow-up tasks.
